SpatialGDK/Source/SpatialGDK/Private/Interop/RPCs/CrossServerRPCService.cpp

```diff
@@ -0,0 +1,476 @@
+
+	TOptional<uint32> Slot = SenderState.Alloc.PeekFreeSlot();
+	if (!Slot)
+	{
+		return EPushRPCResult::DropOverflowed;
+	}
+
+	uint64 NewRPCId = SenderState.LastSentRPCId++;
+	uint32 SlotIdx = Slot.GetValue();
+
+	RPCRingBufferDescriptor Descriptor = RPCRingBufferUtils::GetRingBufferDescriptor(ERPCType::CrossServer);
+	uint32 Field = Descriptor.GetRingBufferElementFieldId(ERPCType::CrossServer, SlotIdx + 1);
+
+	Schema_Object* RPCObject = Schema_AddObject(EndpointObject, Field);
+
+	RPCTarget Target(CrossServerRPCInfo(EntityId, NewRPCId));
+	CrossServer::WritePayloadAndCounterpart(EndpointObject, Payload.Payload, Target, SlotIdx);
+
+	Schema_ClearField(EndpointObject, Descriptor.LastSentRPCFieldId);
+	Schema_AddUint64(EndpointObject, Descriptor.LastSentRPCFieldId, SenderState.LastSentRPCId);
+
+	CrossServer::RPCKey RPCKey(Sender.Entity, NewRPCId);
+	CrossServer::SentRPCEntry Entry;
+	Entry.Target = Target;
+	Entry.SourceSlot = SlotIdx;
+
+	SenderState.Mailbox.Add(RPCKey, Entry);
+	SenderState.Alloc.CommitSlot(SlotIdx);
+
+	return EPushRPCResult::Success;
+}
+
+void CrossServerRPCService::AdvanceView()
+{
+	const FSubViewDelta& SubViewDelta = SubView->GetViewDelta();
+	for (const EntityDelta& Delta : SubViewDelta.EntityDeltas)
+	{
+		switch (Delta.Type)
+		{
+		case EntityDelta::UPDATE:
+		{
+			for (const ComponentChange& Change : Delta.ComponentUpdates)
+			{
+				ComponentUpdate(Delta.EntityId, Change.ComponentId, Change.Update);
+			}
+			break;
+		}
+		case EntityDelta::ADD:
+			PopulateDataStore(Delta.EntityId);
+			break;
+		case EntityDelta::REMOVE:
+		case EntityDelta::TEMPORARILY_REMOVED:
+			CrossServerDataStore.Remove(Delta.EntityId);
+			RPCStore->PendingComponentUpdatesToSend.Remove(
+				EntityComponentId(Delta.EntityId, SpatialConstants::CROSSSERVER_SENDER_ENDPOINT_COMPONENT_ID));
+			RPCStore->PendingComponentUpdatesToSend.Remove(
+				EntityComponentId(Delta.EntityId, SpatialConstants::CROSSSERVER_RECEIVER_ACK_ENDPOINT_COMPONENT_ID));
+			if (Delta.Type == EntityDelta::TEMPORARILY_REMOVED)
+			{
+				PopulateDataStore(Delta.EntityId);
+			}
+			break;
+		default:
+			break;
+		}
+	}
+}
+
+void CrossServerRPCService::ProcessChanges()
+{
+	const FSubViewDelta& SubViewDelta = SubView->GetViewDelta();
+	for (const EntityDelta& Delta : SubViewDelta.EntityDeltas)
+	{
+		switch (Delta.Type)
+		{
+		case EntityDelta::UPDATE:
+		{
+			for (const ComponentChange& Change : Delta.ComponentUpdates)
+			{
+				ProcessComponentChange(Delta.EntityId, Change.ComponentId);
+			}
+			break;
+		}
+		case EntityDelta::ADD:
+			EntityAdded(Delta.EntityId);
+			break;
+		case EntityDelta::REMOVE:
+
+			break;
+		case EntityDelta::TEMPORARILY_REMOVED:
+			EntityAdded(Delta.EntityId);
+			break;
+		default:
+			break;
+		}
+	}
+}
+
+void CrossServerRPCService::EntityAdded(const Worker_EntityId EntityId)
+{
+	for (const Worker_ComponentId ComponentId : SubView->GetView()[EntityId].Authority)
+	{
+		if (!IsCrossServerEndpoint(ComponentId))
```

Comment:
The authority table contains component set ids iirc (there was a similar fix required for on-add rpc execution in the client/server service)

Reply:
Nice catch!